roachtest: pgjdbc failed [need REPLICATION privilege]

roachtest.pgjdbc [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10967169?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10967169?buildTab=artifacts#/pgjdbc) on master @ [06b5ba7888339ed53477ff73f9f3ae77a708ee17](https://github.com/cockroachdb/cockroach/commits/06b5ba7888339ed53477ff73f9f3ae77a708ee17):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-06b5ba7888339ed53477ff73f9f3ae77a708ee17
Tests run against pgjdbc REL42.3.3
5789 Total Tests Run
5055 tests passed
734 tests failed
70 tests skipped
183 tests ignored
0 tests passed unexpectedly
5 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: org.postgresql.replication.ReplicationSlotTest.testCreateTemporaryLogicalSlotPgLowerThan10 - unknown (unexpected)
--- FAIL: org.postgresql.replication.ReplicationSlotTest.testNotAvailableCreatePhysicalSlotWithoutSlotName - unknown (unexpected)
--- FAIL: org.postgresql.replication.ReplicationSlotTest.testCreateTemporaryPhysicalSlotPgLowerThan10 - unknown (unexpected)
--- FAIL: org.postgresql.replication.ReplicationSlotTest.testNotAvailableCreateLogicalSlotWithoutSlotName - unknown (unexpected)
--- FAIL: org.postgresql.replication.ReplicationSlotTest.testNotAvailableCreateLogicalSlotWithoutOutputPlugin - unknown (unexpected)
For a full summary look at the pgjdbc artifacts 
An updated blocklist (pgjdbcBlocklist) is available in the artifacts' pgjdbc log
test artifacts and logs in: /artifacts/pgjdbc/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*pgjdbc.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29942

--- a/pkg/ccl/serverccl/role_authentication_test.go
+++ b/pkg/ccl/serverccl/role_authentication_test.go
@@ -101,43 +101,266 @@ func TestVerifyPassword(t *testing.T) {
 	require.NoError(t, err)
 
 	for _, tc := range []struct {
-		testName                    string
-		username                    string
-		password                    string
-		isSuperuser                 bool
-		canUseReplicationMode       bool
-		shouldAuthenticateSQL       bool
-		shouldAuthenticateDBConsole bool
+		testName                       string
+		username                       string
+		password                       string
+		isSuperuser                    bool
+		canUseReplicationMode          bool
+		shouldAuthenticateSQL          bool
+		shouldAuthenticateDBConsole    bool
+		shouldHaveReplicateModeEnabled bool
 	}{
-		{"valid login", "azure_diamond", "hunter2", false, false, true, true},
-		{"wrong password", "azure_diamond", "hunter", false, false, false, false},
-		{"empty password", "azure_diamond", "", false, false, false, false},
-		{"wrong emoji password", "azure_diamond", "🍦", false, false, false, false},
-		{"correct password with suffix should fail", "azure_diamond", "hunter2345", false, false, false, false},
-		{"correct password with prefix should fail", "azure_diamond", "shunter2", false, false, false, false},
-		{"wrong password all numeric", "azure_diamond", "12345", false, false, false, false},
-		{"wrong password all stars", "azure_diamond", "*******", false, false, false, false},
-		{"valid login numeric password", "druidia", "12345", false, false, true, true},
-		{"wrong password matching other user", "druidia", "hunter2", false, false, false, false},
-		{"root with empty password should fail", "root", "", true, true, false, false},
-		{"some_admin with empty password should fail", "some_admin", "", true, false, false, false},
-		{"empty username and password should fail", "", "", false, false, false, false},
-		{"username does not exist should fail", "doesntexist", "zxcvbn", false, false, false, false},
+		{
+			testName:                       "valid login",
+			username:                       "azure_diamond",
+			password:                       "hunter2",
+			shouldAuthenticateSQL:          true,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "wrong password",
+			username:                       "azure_diamond",
+			password:                       "hunter",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "empty password",
+			username:                       "azure_diamond",
+			password:                       "",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "wrong emoji password",
+			username:                       "azure_diamond",
+			password:                       "🍦",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
 
-		{"user with NOLOGIN role option should fail", "richardc", "12345", false, false, false, false},
+		{
+			testName:                       "correct password with suffix should fail",
+			username:                       "azure_diamond",
+			password:                       "hunter2345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "correct password with prefix should fail",
+			username:                       "azure_diamond",
+			password:                       "shunter2",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+
+		{
+			testName:                       "wrong password all numeric",
+			username:                       "azure_diamond",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "wrong password all stars",
+			username:                       "azure_diamond",
+			password:                       "*******",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "valid login numeric password",
+			username:                       "druidia",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          true,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "wrong password matching other user",
+			username:                       "druidia",
+			password:                       "hunter2",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "root with empty password should fail",
+			username:                       "root",
+			password:                       "",
+			isSuperuser:                    true,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: true,
+		},
+		{
+			testName:                       "some_admin with empty password should fail",
+			username:                       "some_admin",
+			password:                       "",
+			isSuperuser:                    true,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: true,
+		},
+		{
+			testName:                       "empty username and password should fail",
+			username:                       "",
+			password:                       "",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "username does not exist should fail",
+			username:                       "doesntexist",
+			password:                       "zxcvbn",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with NOLOGIN role option should fail",
+			username:                       "richardc",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
 		// The NOSQLLOGIN cases are the only cases where SQL and DB Console login outcomes differ.
-		{"user with NOSQLLOGIN role option should fail SQL but succeed on DB Console", "richardc2", "12345", false, false, false, true},
-		{"user with NOSQLLOGIN global privilege should fail SQL but succeed on DB Console", "has_global_nosqlogin", "12345", false, false, false, true},
-		{"user who inherits NOSQLLOGIN global privilege should fail SQL but succeed on DB Console", "inherits_global_nosqlogin", "12345", false, false, false, true},
-
-		{"replication user should succeed", "replication_user", "12345", false, true, true, true},
-
-		{"user with VALID UNTIL before the Unix epoch should fail", "before_epoch", "12345", false, false, false, false},
-		{"user with VALID UNTIL at Unix epoch should fail", "epoch", "12345", false, false, false, false},
-		{"user with VALID UNTIL future date should succeed", "cockroach", "12345", false, false, true, true},
-		{"user with VALID UNTIL 10 minutes ago should fail", "toolate", "12345", false, false, false, false},
-		{"user with VALID UNTIL future time in Shanghai time zone should succeed", "timelord", "12345", false, false, true, true},
-		{"user with VALID UNTIL NULL should succeed", "cthon98", "12345", false, false, true, true},
+		{
+			testName:                       "user with NOSQLLOGIN role option should fail SQL but succeed on DB Console",
+			username:                       "richardc2",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with NOSQLLOGIN global privilege should fail SQL but succeed on DB Console",
+			username:                       "has_global_nosqlogin",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user who inherits NOSQLLOGIN global privilege should fail SQL but succeed on DB Console",
+			username:                       "inherits_global_nosqlogin",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "replication user should succeed",
+			username:                       "replication_user",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          true,
+			shouldAuthenticateSQL:          true,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: true,
+		},
+		{
+			testName:                       "user with VALID UNTIL before the Unix epoch should fail",
+			username:                       "before_epoch",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with VALID UNTIL at Unix epoch should fail",
+			username:                       "epoch",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with VALID UNTIL future date should succeed",
+			username:                       "cockroach",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          true,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with VALID UNTIL 10 minutes ago should fail",
+			username:                       "toolate",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          false,
+			shouldAuthenticateDBConsole:    false,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with VALID UNTIL future time in Shanghai time zone should succeed",
+			username:                       "timelord",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          true,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
+		{
+			testName:                       "user with VALID UNTIL NULL should succeed",
+			username:                       "cthon98",
+			password:                       "12345",
+			isSuperuser:                    false,
+			canUseReplicationMode:          false,
+			shouldAuthenticateSQL:          true,
+			shouldAuthenticateDBConsole:    true,
+			shouldHaveReplicateModeEnabled: false,
+		},
 	} {
 		t.Run(tc.testName, func(t *testing.T) {
 			execCfg := ts.ExecutorConfig().(sql.ExecutorConfig)
@@ -214,7 +437,7 @@ func TestVerifyPassword(t *testing.T) {
 				)
 			}
 			require.Equal(t, tc.isSuperuser, isSuperuser)
-			require.Equal(t, tc.canUseReplicationMode, canUseReplicationMode)
+			require.Equal(t, tc.shouldHaveReplicateModeEnabled, canUseReplicationMode)
 		})
 	}
 }

--- a/pkg/sql/pgrepl/connect_test.go
+++ b/pkg/sql/pgrepl/connect_test.go
@@ -47,8 +47,9 @@ func TestReplicationConnect(t *testing.T) {
 		replicationMode    string
 		expectedSessionVar string
 		useRoot            bool
+		hasAdmin           bool
 		hasPrivilege       bool
-		hasGrant           bool
+		hasSystemGrant     bool
 		expectErrorCode    pgcode.Code
 	}{
 		{replicationMode: "", hasPrivilege: true, expectedSessionVar: "off"},
@@ -59,8 +60,11 @@ func TestReplicationConnect(t *testing.T) {
 		{replicationMode: "database", hasPrivilege: true, expectedSessionVar: "database"},
 		{replicationMode: "asdf", hasPrivilege: true, expectErrorCode: pgcode.InvalidParameterValue},
 
-		{replicationMode: "true", hasGrant: true, expectedSessionVar: "on"},
-		{replicationMode: "database", hasGrant: true, expectedSessionVar: "database"},
+		{replicationMode: "true", hasSystemGrant: true, expectedSessionVar: "on"},
+		{replicationMode: "database", hasSystemGrant: true, expectedSessionVar: "database"},
+
+		{replicationMode: "true", hasAdmin: true, expectedSessionVar: "on"},
+		{replicationMode: "database", hasAdmin: true, expectedSessionVar: "database"},
 
 		{replicationMode: "true", useRoot: true, expectedSessionVar: "on"},
 		{replicationMode: "database", useRoot: true, expectedSessionVar: "database"},
@@ -71,7 +75,7 @@ func TestReplicationConnect(t *testing.T) {
 		{replicationMode: "database", expectErrorCode: pgcode.InsufficientPrivilege},
 		{replicationMode: "asdf", expectErrorCode: pgcode.InvalidParameterValue},
 	} {
-		t.Run(fmt.Sprintf("hasPrivilege=%t/useRoot=%t/hasGrant=%t/replicationMode=%s", tc.hasPrivilege, tc.useRoot, tc.hasGrant, tc.replicationMode), func(t *testing.T) {
+		t.Run(fmt.Sprintf("hasPrivilege=%t/useRoot=%t/hasSystemGrant=%t/hasAdmin=%t/replicationMode=%s", tc.hasPrivilege, tc.useRoot, tc.hasSystemGrant, tc.hasAdmin, tc.replicationMode), func(t *testing.T) {
 			sqlDB.Exec(t, `DROP USER IF EXISTS testuser`)
 			sqlDB.Exec(t, `CREATE USER testuser LOGIN`)
 
@@ -81,8 +85,11 @@ func TestReplicationConnect(t *testing.T) {
 				sqlDB.Exec(t, `ALTER USER testuser NOREPLICATION`)
 			}
 
-			if tc.hasGrant {
+			if tc.hasSystemGrant {
 				sqlDB.Exec(t, `GRANT replicationsystempriv TO testuser`)
+			}
+			if tc.hasAdmin {
+				sqlDB.Exec(t, `GRANT ADMIN TO testuser`)
 			}
 
 			u := url.User(username.TestUser)

--- a/pkg/sql/pgwire/auth.go
+++ b/pkg/sql/pgwire/auth.go
@@ -220,7 +220,7 @@ func (c *conn) handleAuthentication(
 				ctx,
 				pgerror.Newf(
 					pgcode.InsufficientPrivilege,
-					"must have REPLICATION on the user to start streaming replication",
+					"must be superuser or have REPLICATION role to start streaming replication",
 				),
 			)
 		}

--- a/pkg/sql/user.go
+++ b/pkg/sql/user.go
@@ -163,7 +163,7 @@ func GetUserSessionInitInfo(
 
 			// Only check for replication if the user can login.
 			if canLoginSQL {
-				canUseReplicationMode = authInfo.CanUseReplicationRoleOpt
+				canUseReplicationMode = authInfo.CanUseReplicationRoleOpt || isSuperuser
 				// Only check the global privilege if we do not already have replication
 				// privileges.
 				if !canUseReplicationMode {

